cEP-0017.md

```diff
@@ -0,0 +1,176 @@
+running pip install on bears and then checking their dependencies. However, the
+design of the `cib` tool needs to be able to handle other special cases such as of
+bears that have different installation instructions for each distribution such as
+`DartLintBear` and bears that have bear dependencies.
+
+Additionally, the tool should also be able to provide verbose output, ask user for
+permission before proceeding and install bears related to a specific language.
+
+## Proposed Change
+
+### Here is the procedure for the installation of the bears
+
+1. Identify the valid bear names that need to be installed as specified by the
+   user on the command line.
+2. Install bear on the valid bear names. We will run commands from `CondaRequirement`
+   to check whether the package is uploaded to `anaconda.org` and then check whether
+   it can be installed from `PyPi` using `PipRequirement`.
+   - This will require to create conda packages for bears that have different
+     installation instructions for each distribution. This is done by making `bld.bat`
+     file that has installation instruction for Windows and similarly a `bld.sh` file
+     that has installation instruction for Linux.
+3. After installing the bear we proceed to installing its requirements. We call
+   methods from the requirement classes to do this.
+   - Add new Requirement classes to support different dependencies from various linters.
+   - The DistributionRequirement class needs to be split into different classes for each
+     package manager and call its method to introduct more consistency.
+4. We then proceed to install bear dependencies such as for `ClangCloneDetectionBear` by
+   running `install_bears` method on each dependency bear.
+
+### Here is the process for upgrading bears
+
+1. Check the bear name is installed using `PipRequirement` or `CondaRequirement`.
+2. Iterate through `Requirement` instances to make sure that the required version of the
+   linter is installed or upgrade the version if required.
+
+### Split `DistributionRequirement` for each distribution
+
+Specifically this will cover the following platforms:
+
+| Operating System | Tools      |
+| ---------------- | ---------- |
+| Debian           | apt-get    |
+| Ubuntu           | apt-get    |
+| CentOS           | yum        |
+| Fedora           | dnf        |
+| Arch Linux       | pacman     |
+| openSUSE         | zypper     |
+| Void Linux       | xbps       |
+| macOS            | brew       |
+| Windows          | Chocolatey |
+
+### Using features from depman
+
+Depman provides a way to install dependencies by specifying them through `requirements.yml`.
+Currently, only `pip` and `apt-get` are supported but by adding support for new installers we
+would be able to generate a `requirements.yml` specifying the bears and their dependencies.
+It would be useful to have `requirements.yml` in each project for the user that uses coala
+therefore, the cib should provide this additional option.
```

Comment:
Having a dependency manager that invokes another dependency manager to solve a subset of problems seems to me at best to be a temporary solution, and bad design in the long run.  In my opinion, one program or the other should do the heavy lifting of satisfying a set of dependencies.  As I see it, the only good options for integrating the projects are to either (1) integrate the appropriate functionality from `depman` into `cib` and/or `dependency_management`, or (2) replace `dependency_management` requirement classes with `depman` `Dependency` subclasses.

While it is not currently documented, there is a `depman` API that will allow for requirements to be specified using [`depman.dependency.Dependency`](https://github.com/mbodenhamer/depman/blob/master/depman/dependency.py#L57) subclasses in lieu of composing or generating `requirements.yml`.  Such usage is illustrated in the unit tests, such as [here](https://github.com/mbodenhamer/depman/blob/76d76eb77f9aacbc75ed2dcce5137bb910465c49/depman/tests/test_dependency.py#L119).  As one can see, the call signature for `Pip` is practically identical to that of `PipRequirement`.  As such, if `depman` classes were to be used instead of `dependency_management` classes in bear `REQUIREMENTS`, only the import statement at the top of each bear module would need to be changed.

I'm not committed to any particular implementation strategy and only offer the above as an example.  However, if `depman` is to be used as a means of resolving bear dependencies, then I think that things will work out better if those dependencies are specified in the bear definitions themselves.  Generating additional files should be avoided if possible, particularly when the necessary functionality can be accessed from within Python.

Reply:
Thanks for the feedback. 
> Having a dependency manager that invokes another dependency manager to solve a subset of problems seems to me at best to be a temporary solution, and bad design in the long run. In my opinion, one program or the other should do the heavy lifting of satisfying a set of dependencies.

I agree.

> However, if depman is to be used as a means of resolving bear dependencies, then I think that things will work out better if those dependencies are specified in the bear definitions themselves. 

This would mean adding all the required package managers that we currently support to `depman`. But I am not sure about the benefit we would have over the current `REQUIREMENT` classes that we have.

Reply:
> This would mean adding all the required package managers that we currently support to depman and I am not sure if there would be any considerable advantage that `depman` classes would have over the current `REQUIREMENT` classes in `dependency_management`.

This is the crux of the issue.  The primary advantage that `depman` provides is context-awareness.  If there is a need to create two different sets of requirements for two different system configurations that share a common set of requirements, that is straightforward to do and the logic is built into the program.  However, I don't think that implementing this kind of feature in `dependency_management` / `cib` would be too complicated to do.

The other two advantages that `depman` provides are (1) package manager call optimization (minimizing the number of times that the package manager is invoked on the system), and (2) support for downgrading to a previous package version.  Implementing call optimization (1) would not be nearly as straightforward as implementing context-awareness, but it is a less important feature in the grand scheme of things.  Cases where (2) is necessary are rare, and supporting downgrades does not appear to be a supported/intended feature of `dependency_management`/`cib` anyway.

With all of that said, I think the easiest and most cost-effective path forward is to implement the needed features in `cib` / `dependency_management`, particularly context-awareness.  A straightforward first step after breaking up `DistributionRequirement` into different distribution-specific requirement classes would be to (1) implement a method in the distribution specific classes that checks whether or not the user is running the target distribution, and (2) modify `cib` so that it only tries to satsify those requirements that pass the distribution check.  That would allow multiple distribution-specific requirements to be specified for each bear, but `cib` would only attempt to satisfy those requirements that match the user's distribution (or requirements that are distribution-agnostic).